core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -260,6 +261,64 @@ public UpdateSchema updateColumnDoc(String name, String doc) {
     return this;
   }
 
+  @Override
+  public UpdateSchema updateSchema(Schema newSchema) {
+    TypeUtil.visit(newSchema, new ApplyUpdates(schema, this));
+    return this;
+  }
+
+  private static final Joiner DOT = Joiner.on(".");
+
+  private static class ApplyUpdates extends TypeUtil.SchemaVisitor<Void> {
+    private final Schema schema;
+    private final UpdateSchema updateSchema;
+    private final Map<String, Integer> indexByName;
+
+    private ApplyUpdates(Schema schema, UpdateSchema updateSchema) {
```

Comment:
It would be nice to update the names to signal a bit better what's happening here. From the usage, it looks like the schema passed in here is the current table schema and the schema that is traversed is the updated schema. Maybe some Javadoc for this visitor class would help.